osquery/tables/system/windows/ads.cpp

```diff
@@ -0,0 +1,179 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <boost/algorithm/string/trim.hpp>
+#include <boost/filesystem.hpp>
+#include <string>
+#include <windows.h>
+
+#include <osquery/core/tables.h>
+#include <osquery/filesystem/filesystem.h>
+#include <osquery/logger/logger.h>
+#include <osquery/utils/base64.h>
+#include <osquery/utils/chars.h>
+#include <osquery/utils/conversions/split.h>
+#include <osquery/utils/conversions/windows/strings.h>
+#include <osquery/utils/scope_guard.h>
+
+namespace fs = boost::filesystem;
+
+namespace osquery {
+namespace tables {
+
+const std::string kZoneIdentifierKey = "Zone.Identifier";
+
+void setRow(QueryData& results,
+            const std::string& path,
+            const std::string& key,
+            const std::string& value) {
+  Row r;
+  r["path"] = path;
+  r["directory"] = boost::filesystem::path(path).parent_path().string();
+  r["key"] = key;
+  auto value_printable = isPrintable(value);
+  r["value"] = value_printable ? value : base64::encode(value);
+  r["base64"] = value_printable ? INTEGER(0) : INTEGER(1);
+  results.push_back(r);
+}
+
+void parseZoneIdentifier(QueryData& results,
+                         const std::string& path,
+                         const std::string& streamData) {
+  auto lines = split(streamData, "\n");
+  for (const auto& line : lines) {
+    auto key_len = line.find_first_of("=");
+    if (key_len == std::string::npos) {
+      continue;
+    }
+
+    setRow(results,
+           path,
+           line.substr(0, key_len),
+           line.substr(key_len + 1, line.size()));
+  }
+}
+
+// Process a file and extract all stream names and data.
+void enumerateStreams(QueryData& results, const std::string& path) {
+  WIN32_FIND_STREAM_DATA findStreamData;
+  HANDLE hFind = FindFirstStreamW(stringToWstring(path).c_str(),
+                                  FindStreamInfoStandard,
+                                  &findStreamData,
+                                  0);
+
+  auto fd_guard = scope_guard::create([&] { FindClose(hFind); });
+
+  bool isFirstStream = true;
+
+  if (hFind != INVALID_HANDLE_VALUE) {
+    do {
+      // Skip first stream

```

Comment:
Why? As a reader previously unfamiliar with ADS, I don't understand why the first stream is not relevant. Is it because the first stream is the standard file?

Reply:
Yes, per [FindFirstStreamW docs](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findfirststreamw#remarks):
> The FindFirstStreamW function opens a search handle and returns information about the first $DATA stream in the specified file or directory. For files, this is always the default, unnamed data stream, "::$DATA". Directories do not have $DATA streams by default and cannot have an unnamed data stream, but may have named data streams set after they have been created.

The unnamed data stream is the file contents so I thought it's best to skip that one. Although I have just noticed that by the way the table works now it will skip the first stream of a directory when it shouldn't do it since they can't have unnamed data streams.

I will address this alongside the other comments.